Frameworks/Foundation/NSOrderedSet.mm

```diff
@@ -14,311 +14,403 @@
 - (instancetype)initWithSet:(NSSet*)set copyItems:(BOOL)flag {
-    UNIMPLEMENTED();
-    return StubReturn();
+    if (set == nil) {
+        return [self init];
+    }
+
+    NSMutableArray* arrayObjs = [NSMutableArray array];
+
+    for (id item : set) {
+        [arrayObjs addObject:item];
+    }
+    return [self initWithArray:arrayObjs range:NSMakeRange(0, [arrayObjs count]) copyItems:flag];
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (instancetype)init {
-    UNIMPLEMENTED();
-    return StubReturn();
+    return [super init];
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (BOOL)containsObject:(id)object {
-    UNIMPLEMENTED();
-    return StubReturn();
+    return mapContainer.find(object) != mapContainer.end();
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (void)enumerateObjectsAtIndexes:(NSIndexSet*)indexSet
                           options:(NSEnumerationOptions)opts
                        usingBlock:(void (^)(id, NSUInteger, BOOL*))block {
-    UNIMPLEMENTED();
+    if (block == nil) {
+        [NSException raise:NSInvalidArgumentException format:@"block can not be nil."];
+    }
+
+    [indexSet enumerateIndexesWithOptions:opts
+                               usingBlock:^(NSUInteger idx, BOOL* stop) {
+                                   block([self objectAtIndex:idx], idx, stop);
+                               }];
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (void)enumerateObjectsUsingBlock:(void (^)(id, NSUInteger, BOOL*))block {
-    UNIMPLEMENTED();
+    return [[self array] enumerateObjectsUsingBlock:block];
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (void)enumerateObjectsWithOptions:(NSEnumerationOptions)opts usingBlock:(void (^)(id, NSUInteger, BOOL*))block {
-    UNIMPLEMENTED();
+    return [[self array] enumerateObjectsWithOptions:opts usingBlock:block];
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (id)objectAtIndex:(NSUInteger)index {
-    UNIMPLEMENTED();
-    return StubReturn();
+    if (index >= [self count]) {
+        [NSException raise:NSRangeException format:@"index is greater than the orderedSet size %u", [self count]];
+    }
+    return setContainer[index];
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
+*/
+- (id)firstObject {
+    return [self objectAtIndex:0];
+}
+
+/**
+ @Status Interoperable
+*/
+- (id)lastObject {
+    return [self objectAtIndex:([self count] - 1)];
+}
+
+/**
+ @Status Interoperable
 */
 - (id)objectAtIndexedSubscript:(NSUInteger)index {
-    UNIMPLEMENTED();
-    return StubReturn();
+    return [self objectAtIndex:index];
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (NSArray*)objectsAtIndexes:(NSIndexSet*)indexes {
-    UNIMPLEMENTED();
-    return StubReturn();
+    if (indexes == nil) {
+        [NSException raise:NSInvalidArgumentException format:@"indexes can not be nil."];
+    }
+
+    NSMutableArray* ret = [NSMutableArray array];
```

Comment:
never mind, it's actually not necessary here. see [https://developer.apple.com/library/ios/documentation/Cocoa/Conceptual/Blocks/Articles/bxVariables.html#//apple_ref/doc/uid/TP40007502-CH6-SW6](https://developer.apple.com/library/ios/documentation/Cocoa/Conceptual/Blocks/Articles/bxVariables.html#//apple_ref/doc/uid/TP40007502-CH6-SW6) if you're curious though

---
In reply to: [62386471](https://github.com/Microsoft/WinObjC/pull/482#discussion_r62386471) [](ancestors = 62386471)